subprojects/language-native/src/integTest/groovy/org/gradle/language/cpp/CppStaticLibraryPublishingIntegrationTest.groovy

```diff
@@ -0,0 +1,196 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.language.cpp
+
+import org.gradle.nativeplatform.fixtures.ToolChainRequirement
+import org.gradle.nativeplatform.fixtures.app.CppAppWithLibrariesWithApiDependencies
+import org.gradle.test.fixtures.file.TestFile
+import org.junit.Assume
+
+import static org.gradle.api.platform.MachineArchitecture.*
+import static org.gradle.api.platform.OperatingSystemFamily.*
+
+class CppStaticLibraryPublishingIntegrationTest extends AbstractCppPublishingIntegrationTest {
+    def "can publish a library and its dependencies to a Maven repository when multiple target operating systems are specified"() {
+        def app = new CppAppWithLibrariesWithApiDependencies()
+        def targetMachines = [machine(WINDOWS, currentArchitecture), machine(LINUX, currentArchitecture), machine(MACOS, currentArchitecture)]
+
+        given:
+        settingsFile << "include 'deck', 'card', 'shuffle'"
+        buildFile << """
+            subprojects {
+                apply plugin: 'cpp-library'
+                apply plugin: 'maven-publish'
+                
+                group = 'some.group'
+                version = '1.2'
+                publishing {
+                    repositories { maven { url '${mavenRepo.uri}' } }
+                }
+                
+                components.withType(CppComponent) {
+                    linkage = [Linkage.STATIC]
+                    targetMachines = [machines.windows().architecture('${currentArchitecture}'), machines.linux().architecture('${currentArchitecture}'), machines.macOS().architecture('${currentArchitecture}')]
+                }
+            }
+            project(':deck') { 
+                dependencies {
+                    api project(':card')
+                    implementation project(':shuffle')
+                }
+            }
+        """
+        app.deck.writeToProject(file('deck'))
+        app.card.writeToProject(file('card'))
+        app.shuffle.writeToProject(file('shuffle'))
+
+        when:
+        run('publish')
+
+        then:
+        assertMainModuleIsPublished('some.group', 'deck', '1.2', targetMachines, ["some.group:card:1.2"])
+        assertVariantsArePublished('some.group', 'deck', '1.2', ['debug', 'release'], targetMachines.findAll { it.os == currentOsFamilyName }, ["some.group:shuffle:1.2", "some.group:card:1.2"])
+
+        and:
+        assertMainModuleIsPublished('some.group', 'card', '1.2', targetMachines)
+        assertVariantsArePublished('some.group', 'card', '1.2', ['debug', 'release'], targetMachines.findAll { it.os == currentOsFamilyName })
+
+        and:
+        assertMainModuleIsPublished('some.group', 'shuffle', '1.2', targetMachines)
+        assertVariantsArePublished('some.group', 'shuffle', '1.2', ['debug', 'release'], targetMachines.findAll { it.os == currentOsFamilyName })
```

Comment:
Perhaps move the `targetMachines.findAll { }` expression to a helper method, as this expression appears in loads of places and it would be clearer to give the expression an expressive name (by moving it to a method).